types/trigger.d.ts

```diff
@@ -154,9 +159,19 @@ export type TriggerSet<Data extends RaidbossData> = {
   timelineTriggers?: (TimelineTrigger<Data> | DisabledTrigger)[];
   timelineReplace?: TimelineReplacement[];
   timelineStyles?: TimelineStyle[];
-  initData?: DataInitializeFunc<Data>;
 }
 
+// If Data contains required properties that are not on RaidbossData, require initData
+export type TriggerSet<Data extends RaidbossData> = BaseTriggerSet<Data> &
+  (Required<RaidbossData> extends Required<Data> ? Required<Data> extends Required<RaidbossData> ?
+  {
+    initData?: DataInitializeFunc<Data>;
+  } : {
+    initData: DataInitializeFunc<Data>;
+  } : {
+    initData: DataInitializeFunc<Data>;
+  });
```

Comment:
I think you have an extra ternary.  My understanding of `extends` is that `X extends Y` means X can be assigned to Y.  In this case, the types are unions, and so really you just want to check that `Data` does not have any extra properties that `RaidbossData` doesn't have.  `export type TriggerSet<Data extends RaidbossData>` already verifies that `Data` has all the properties that `RaidbossData` has, I think?

```suggestion
export type TriggerSet<Data extends RaidbossData> = BaseTriggerSet<Data> &
  (Required<Data> extends Required<RaidbossData> ?
  {
    initData?: DataInitializeFunc<Data>;
  } : {
    initData: DataInitializeFunc<Data>;
  });
```

Reply:
Yeah, this makes sense. I think it might also be a good idea to store `type RequiredRaidbossDataFields = Required<RaidbossData>` as well, since this could cause some compile-time performance impact once a ton of triggers are converted?

Reply:
I do not know enough about tsc to know if "storing" saves work here.  My instinct is that this is probably premature optimization, especially for a once-per-triggerset sort of type.